Improve CPU load calculation

LeandroMartinsdS opened the pull request:
Improve CPU load calculation for Power PMAC, being possible to read the usage for up to 4 cores.
For ARM architecture, supports custom core management reading the core that executes each task (phase, servo, RT and background).

LeandroMartinsdS committed:
Remove BgSleepTime from CPU calculation

The PowerPMAC firmware already adjusts BgDeltaTime when BgSleepTime changes

Remove repeated snippet caused by merging with conflicts

LeandroMartinsdS committed:
Add support for custom tasks allocation in CPU load calculation

Gets the number of cores for each type of CPU.

Defines in which core the tasks are executed - for ARM arch it reads
from the controller.
It doesn't take in account Ethercat tasks.

Calculates the CPU load for each core based on the tasks defined -
maximum of 4 cores.

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -222,7 +222,7 @@ pmacController::pmacController(const char *portName, const char *lowLevelPortNam
   csResetAllDemands = false;
   csCount = 0;
   Sys_CPUFreq_ = 0;
-  Sys_CPUType_ = 0;
```
I think this line can be deleted rather than just commented out, especially as the code where it was used has been deleted

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -4150,9 +4165,109 @@ asynStatus pmacController::readDeviceType() {
+    } else if (strcmp(cpu_.c_str(),"PowerPC,APM86xxx") == 0) {
+      cpuNumCores_ = 2;
+    } else if (strcmp(cpu_.c_str(),"arm,LS1021A") == 0) {
+      strcpy(cmd, "Sys.Cores");
```
Can Sys.Cores be run when the CPU type is first established, and used to determine the number of cores for all CPU models rather than just the ARM?

LeandroMartinsdS replied:
The Sys.Cores was added in V2.6 firmware, that is also the first version that supports the ARM arch.

On second thought, it might not be necessary to use Sys.Cores, since the LS1021A has a dual-core processor.
I'll try to confirm the cpu response for a quad-core processor.

LeandroMartinsdS replied:
The response for ARM,quad-core should be "arm,LS1043A".

About running Sys.Cores just after the reading the CPU type, that was the first idea, but I had to move to pmac constructor, since I'm using the immediateWriteRead (from pmacMessageBroker) to get the Sys.Cores reply.

As I said, we can hard code the number of cores as 2 for "arm,LS1021A"; and 4 for "arm,LS1043A".

That way, I think it'll be possible to call getCpuNumCores just after readDeviceType in initialSetup.

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -4150,9 +4165,109 @@ asynStatus pmacController::readDeviceType() {
+      status = parseIntegerVariable(cmd, reply,
+                                    "Read the number of CPU cores", cpuNumCores_);
+    } else {
+      debug(DEBUG_ERROR, functionName, "CPU not recognized");
```
This error case doesn't raise an error with asynSuccess meaning that the following setIntegerParam would get set with an uninitialised value

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -4150,9 +4165,109 @@ asynStatus pmacController::readDeviceType() {
   return status;
 }
 
+asynStatus pmacController::getTasksCore() {
```
This function could make use of the getCpuNumCores function to make it model agnostic - ie each if clause could check against the number of cores, rather than the model

LeandroMartinsdS replied:
As the Sys.Cores, the commands to get which core execute each task require V2.6 firmware, making it (indirectly) dependent of the CPU.

Both APM86xxx and LS1021A have 2 cores, just in the latest one is possible to customize in which core each task is executed.

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -4150,9 +4165,109 @@ asynStatus pmacController::readDeviceType() {
+    cpuCoreTasks_[PPMAC_CPU_SERVOTASK] = 1;
+    cpuCoreTasks_[PPMAC_CPU_RTTASK]    = 1;
+
+  // Dual-core ARM
```
Is this comment meant to say dual core?  It looks like it might be quad core?

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -4150,9 +4165,109 @@ asynStatus pmacController::readDeviceType() {
+    debug(DEBUG_TRACE, functionName, "CPU type", cpu_.c_str());
+    debug(DEBUG_TRACE, functionName, "Number of cores", cpuNumCores_);
+
+    for (int task_idx = 0; task_idx < 4; task_idx++) {
```
Could the hard-coded 4 be replaced by cpuNumCores to remove a magic number and make the code more future proof?

LeandroMartinsdS replied:
It actually should be PPMAC_CPU_TASKS_NUM.

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -4150,9 +4165,109 @@ asynStatus pmacController::readDeviceType() {
+        case 3:   strcpy(cmd, "Sys.CoreBackground") ; break;
+      }
+      // get task Core
+      status = pBroker_->immediateWriteRead(cmd, reply);
```
The return status of pBroker_->immediateWriteRead is not checked - could this cause a problem below if it fails?

JamesOHeaDLS commented on pmacApp/src/pmacController.h:
```diff
@@ -212,6 +216,13 @@
 #define PPMAC_CPU_RTI_PERIOD     "Sys.RtIntPeriod"
 #define PPMAC_CPU_BGSLEEP_TIME   "Sys.BgSleepTime"
 
+#define PPMAC_CPU_MAXCORES      4
```
Is PPMAC_CPU_MAXCORES used anywhere?

LeandroMartinsdS replied:
Now being used in getCpuNumCores for checking the number of cores is within the supported range.

JamesOHeaDLS commented on pmacApp/src/pmacController.h:
```diff
@@ -524,7 +547,7 @@ class pmacController
     int csCount;
     int Sys_CPUFreq_;
-    int Sys_CPUType_;
+    // int Sys_CPUType_;
```
Sys_CPUType_ can be deleted

JamesOHeaDLS commented on pmacApp/src/pmacController.h:
```diff
@@ -524,7 +547,7 @@ class pmacController
     int Sys_CPUFreq_;
-    int Sys_CPUType_;
+    // int Sys_CPUType_;
     int Sys_BgSleepTime_;
```
I think Sys_BgSleepTime_ is not used anymore?

LeandroMartinsdS replied:
You're right James, Sys_BgSleepTime_ is not used anymore.

The following aren't being used as well:
  Sys_ServoPeriod_
  Sys_RtIntPeriod_
  Sys_PhaseOverServoPeriod_

Any thoughts about deleting them?

JamesOHeaDLS replied:
I think any variable that is not being used should be deleted, otherwise we build up lots of commented out lines which can be confusing when another developer looks at the code at a later date.

Also, the change tracking from git means that we always have a record of what was there if we ever need to check anything later

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -4150,9 +4165,109 @@ asynStatus pmacController::readDeviceType() {
+      if (status == asynSuccess) {
+        cpuCoreTasks_[task_idx] = taskCore;
+      } else {
+        debugf(DEBUG_ERROR, functionName,
```
This debug message could state the Sys command which had the issue rather than the task index number

JamesOHeaDLS reviewed (commented):
Hi @LeandroMartinsdS 

Overall it looks good.  I have added some inline review comments to the code.

Have you tested how the system behaves if no CPU is recognised?

Thanks,

James

LeandroMartinsdS commented:
I made this test
`Have you tested how the system behaves if no CPU is recognised?`

Now, the usage percentage of each task is summed into the first core usage.
I can easily change that to set all of the cores to 0%  -  I'm not sure if that would be clear enough to indicate an error.

LeandroMartinsdS committed:
Remove unused variables and create defined macros

LeandroMartinsdS committed:
Include severity for CPUNUMCORES record

LeandroMartinsdS committed:
Update pmacController.edl for multi-core CPUs

LeandroMartinsdS committed:
Fix debug message

LeandroMartinsdS commented:
@JamesOHeaDLS, I think that all of your suggestions are sorted.

As I have made a few changes, would be nice to have a look again just in case I've missed something.

Thanks.

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -4189,9 +4161,128 @@ asynStatus pmacController::readDeviceType() {
+  if(status == asynSuccess) {
+    if(cpuNumCores_ < 1 || cpuNumCores_ > PPMAC_CPU_MAXCORES) {
+      debugf(DEBUG_ERROR, functionName,
+            "Detected number of CPU Cores (%d) is out of supported range (1 to 4)", cpuNumCores_);
```
This string can use the PPMAC_CPU_MAXCORES it is evaluating against instead of '4'

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -4189,9 +4161,128 @@ asynStatus pmacController::readDeviceType() {
+  if (strcmp(cpu_.c_str(), "PowerPC,460EX") == 0) {
+    debug(DEBUG_TRACE, functionName, "CPU type", cpu_.c_str());
+    debug(DEBUG_TRACE, functionName, "Number of cores", cpuNumCores_);
+
```
The two lines
```
debug(DEBUG_TRACE, functionName, "CPU type", cpu_.c_str());
debug(DEBUG_TRACE, functionName, "Number of cores", cpuNumCores_);
```
are repeated multiple times within this if clause, but are not dependent on anything inside it.  So I think they can be taken outside of the clause and called just once each

LeandroMartinsdS replied:
Moved to `getCpuNumCores` method, as it makes more sense to be there.
Also substitutes `debug` for `printf`, since it's only executed at the constructor, where the `DEBUG_TRACE` option is not already set.

JamesOHeaDLS commented on pmacApp/Db/pmacController.template:
```diff
@@ -229,13 +229,70 @@ record(mbbi, "$(P):PMACTYPE") {
+  field(DTYP, "asynInt32")
+  field(SCAN, "I/O Intr")
+  field(INP, "@asyn($(PORT),0)PMAC_C_CPU_NUM_CORES")
+  field(HIGH, "4")
```
Do we want severity on this record?  Wouldn't it always be in an alarm state if the alarm is based on number of CPU cores?

LeandroMartinsdS replied:
The alarms were implemented to detect at the EPICS level if reading the number of cores was not possible, or if the number of cores exceeded the currently supported limit of 4 cores. However, upon further reflection, it seems more appropriate to allow the driver to handle this entirely.

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -1843,68 +1820,63 @@ asynStatus pmacController::fastUpdate(pmacCommandStore *sPtr) {
+        int core = cpuCoreTasks_[task_idx];
+        if (core >= 0) {
+          cpuLoad_[core] += tasksPercent[task_idx];
+        }
```
cpuLoad_[core] is not initialised on line 1787 above
The if statement of 
`if(core >= 0) `
means that it is possible for an element of cpuLoad_ to remain uninitialised

The for loop below reads out from every element and so could try to read an uninitialised value

LeandroMartinsdS replied:
cpuLoad_ initialised as vector.

JamesOHeaDLS reviewed (commented):
Looks good overall.  I have made a couple of comments/suggestions

LeandroMartinsdS committed:
Revert LF removal from cpu command reply

PR https://github.com/dls-controls/pmac/pull/103 sorts that in the correct way

LeandroMartinsdS committed:
Initilize cpuLoad_ as vector

LeandroMartinsdS committed:
Update debug message including PPMAC_CPU_MAXCORES

LeandroMartinsdS committed:
Remove alarms from CPUNUMCORES record

LeandroMartinsdS committed:
Update Cpu type and cores debug messages

Move debug messages to getCpuNumCores
Changed from debugf to printf as it runs only during the startup.

LeandroMartinsdS committed:
Include geobrick and clipper in getCpuNumCores

LeandroMartinsdS committed:
Tweak pmacController screen

LeandroMartinsdS commented:
A personal reminder to change the debug level of each task frequency/time/percentage from `DEBUG_TRACE` to `DEBUG_VARIABLE`

LeandroMartinsdS committed:
Change CPU usage debug messages to DEBUG_VARIABLE

LeandroMartinsdS commented:
Tagged as 2-6-4b3.
